### ISSUE TITLE
Doxygen is not showing all called functions in graph plot

### ISSUE THREAD

**Issue by r2com** [doxygen/doxygen#9070]
I have some codebase, inside I am calling some function "D3D12CreateDevice" which is called by returning HRESULT like below:

![image](https://user-images.githubusercontent.com/5223533/163723003-40432758-42ba-4825-857d-ca180f1c679e.png)

But when the plots generated, I get:

![image](https://user-images.githubusercontent.com/5223533/163723028-dd5770d5-8377-43fd-bc26-3c45eae9e8fb.png)

it show HRESULT box, but not the D3D12CreateDevice function

I got all options YES, like CALL_GRAPH, CALLER_GRAPH , HAVE_DOT, EXTRACT_ALL etc.

that function which I'm trying to get reflected in plot is inside d3d12.h file, adding directory of that file as inlcude in Input section does not help either.

any ideas how to achieve the result?

**Comment by albert-github** [comment]
Please don't post images on external sources but copy / paste drag /drop them in the question.
To be able to say something sensible about it:
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.
- Please also specify the **full** doxygen version used (`doxygen -v`).

**Comment by r2com** [comment]
[directx.zip](https://github.com/doxygen/doxygen/files/7967907/directx.zip)
[dxtest5.zip](https://github.com/doxygen/doxygen/files/7967908/dxtest5.zip)
![DIAGRAM_PROBLEM](https://user-images.githubusercontent.com/18331178/151735936-9ad0ce46-6633-41bd-8861-bae26a308d2e.jpg)

it is hard to attach self contained example, but I will try and explain.

I am attaching 2 files:
"dxtest5" is main project folder where stuff compiled; the "src" folder within this folder is shown in "Input" of GUI frontend
"directx" is folder where additional files are used by my projects, and it is Also shown as "Input" in GUI frontend

that's all about it, the rest is simple, the "Doxygen" file is inside "dxtest5" folder.

On top of that, I also attached "DIAGRAM_PROBLEM" picture, which clearly explains WHAT exactly is my problem after successful run.

doxygen version shown in About of GUI frontend is 1.9.3

**Comment by albert-github** [comment]
It looks like the `cpp_quote` of the idl files is the problem in this case. I think that by adding:
```
PREDEFINED += cpp_quote(x)=
MACRO_EXPANSION = YES
```
to your `Doxyfile` the problem may be solved.

Furthermore I see that also `__declspec` gives some problems, so it is advised to also add:
```
PREDEFINED += __declspec(x)=
```

Though this is just a workaround (and will miss some calls).


I've just pushed a proposed patch, pull request #9087

**Comment by r2com** [comment]
Unfortunately adding proposed 3 lines of code to my Doxyfile did not solve it, still just see "HRESULT"

**Comment by albert-github** [comment]
The question is "Where do you see the `HRESULT`"?

With the original 1.9.3 version I see:
![image](https://user-images.githubusercontent.com/5223533/151964261-00158410-e2f2-4526-ad7a-7a115c44fce9.png)

With the 1.9.3 version and adding the mentioned lines to the `Doxyfile`:
```
PREDEFINED += __declspec(x)=
PREDEFINED += cpp_quote(x)=
MACRO_EXPANSION = YES
```
 I see:
![image](https://user-images.githubusercontent.com/5223533/151964150-e6133525-e966-4cc8-9aa5-a1a3e1ea2b64.png)

So with the adjusted `Doxyfile` I dont see the `HRESULT` anymore, though I lost `D3D12SerializeRootSignature()` in the graph.

With the current master plus the proposed patch (#9087) and the lines:
```
PREDEFINED += __declspec(x)=
MACRO_EXPANSION = YES
```
I see:
![image](https://user-images.githubusercontent.com/5223533/151964427-d73e4494-1974-49d6-911c-5cdb38dab05d.png)

So here the `D3D12SerializeRootSignature()` is restored again and I don't see in this graph `HRESULT`.

I also think it is better to add the following settings:
```
PREDEFINED += BEGIN_INTERFACE=
PREDEFINED += END_INTERFACE=
```

**Comment by albert-github** [comment]
One more `PREDEFINED`:
```
PREDEFINED += STDMETHODCALLTYPE=
```

**Comment by r2com** [comment]
Ok, so here is what I have added in total:
```
PREDEFINED += cpp_quote(x)=
MACRO_EXPANSION = YES
PREDEFINED += __declspec(x)=
PREDEFINED += BEGIN_INTERFACE=
PREDEFINED += END_INTERFACE=
PREDEFINED += STDMETHODCALLTYPE=
```

And here is a result, I am interested in D3DApp::InitDirect3D() function:

![BLOCK_GRAPH_ISSUE](https://user-images.githubusercontent.com/18331178/152104373-f16cfc03-ddfc-4958-8658-d1a26c4b6d54.jpg)

And yes, indeed HRESULT is gone, however; problem is:

- I do not have the graph plot view of the D3DCreateDevice function which is called this way: 
```
	HRESULT hardwareResult = D3D12CreateDevice(
		nullptr,             // default adapter
		D3D_FEATURE_LEVEL_12_2,
		IID_PPV_ARGS(&md3dDevice));
```

- I also do not have graph plot view of CreateFence function called this way:
```
	ThrowIfFailed(md3dDevice->CreateFence(0, D3D12_FENCE_FLAG_NONE,
		IID_PPV_ARGS(&mFence)));
```

Also, was I supposed to just add those lines or somehow apply some new patch? Does it mean I must recompile or? because I am using it on Win10 and just installed executable Doxygen. Could it be reason why it kinda worked but still has issues pointed above?

**Comment by albert-github** [comment]
- Executable for Windows 10, Yes you have to have a new executable. Some executables are available for a limited amount of time by means of the GitHub Actions and for this patch this is also the case see https://github.com/doxygen/doxygen/actions/runs/1772673310
- I'm currently looking into some more issues regarding the handling of the `cpp_quote` and some other small issues.
  Hopefully this will also solve some of the other problems you mentioned.

**Comment by r2com** [comment]
thanks but i tried downloading from above link files "[Windows Latest MSVC Release build artifacts](https://github.com/doxygen/doxygen/suites/5121157808/artifacts/153894027)"

but when I unpacked exe to /bin folder of docygen it complained with error about some Qt5 dll.. any idea what am I doing wrong?

**Comment by albert-github** [comment]
I've just pushed an update regarding some problems I saw (executable shoudl appear in https://github.com/doxygen/doxygen/actions/runs/1788662911).

- Regarding the Qt5.dll, looks like a discrepancy between the release build (static) and the Github Action build. Best is indeed to use the "old" doxywizard.
- Regarding the `ThrowIfFailed`, this has to do with the usage of the `define` , this is a known limitation of the doxygen call graph (theer are a number of issues about this problem).

**Comment by r2com** [comment]
Alright, after I used release packaged .exe, I could see some of the functions I did not see before which is good.
At least what I have now is much better than it was before.

So, now you are saying that ThrowIfFailed is a permanent limitation or there is chance that in future updates this will be solved too?

Also, I assume I shouldn't update for next version in order not to loose this patch right? Or in 1.9.5 or something this will be integrated?

in any way, it would be nice if plotter plotted *all* called functions, since plotter is the only tool for analyzing large projects with hundreds of calls, and re-checking each call manually to make sure you did not miss something kinda defeats purpose of plotter. In any case, its better than nothing anyway. thanks for help.

**Comment by albert-github** [comment]
> So, now you are saying that ThrowIfFailed is a permanent limitation or there is chance that in future updates this will be solved too?

I hope that this limitation will be solved one day.

> Also, I assume I shouldn't update for next version in order not to loose this patch right? Or in 1.9.5 or something this will be integrated?

- When the #9087 is integrated in the master it will be in the next release. Depending on the time of the time of the integration this can be 1.9.4, 1.9.5, ...
- When a new release is is present you can update to that version and apply the patch there  (will require your own build).

**Comment by r2com** [comment]
sorry but I thought that at the time of 1.9.5 or 1.9.6 whatever all I'll have to do is just download and install new binary and its gonna be there? i.e. I wont need to do my own build or patching anymore, right?

**Comment by albert-github** [comment]
I think my last sentence:
> When a new release is is present you can update to that version and apply the patch there (will require your own build).

was not 100% clear it should have read:
- When the patch hasn't yet been integrated and a new release is present you can update to that version and apply the patch there (will require your own build).

**Comment by albert-github** [comment]
Some remarks regarding the Qt5 dll "problem" (mentioned in https://github.com/doxygen/doxygen/issues/9070#issuecomment-1028569248 and https://github.com/doxygen/doxygen/issues/9070#issuecomment-1028793432), relevant references to the GitHub Actions are https://github.com/jurplel/install-qt-action/issues/128 and https://github.com/jurplel/install-qt-action/issues/53 / https://github.com/jurplel/install-qt-action/issues/53#issuecomment-675214158.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
